### PR TITLE
Search and filter books

### PR DESCRIPTION
**Pull request by luislopez0690**

### CODE REVIEW

File: app/components/book-filter.js [hunk 1]
```diff
@@ -0,0 +1,18 @@
+import Component from '@ember/component';
+import EmberObject, { computed } from '@ember/object';
+
+export default Component.extend({
+  classNames: ['book-filter'],
+
+  actions: {
+    filterParam(filterProperty) {
+      if (filterProperty == "name") {
+        this.set('currentFilterValue', "Title");
```
Comment (juandsierra82) [inline app/components/book-filter.js]: ''

File: app/components/books-deck.js [hunk 2]
```diff
@@ -1,29 +1,54 @@
...
 import { inject as service } from '@ember/service';
+import EmberObject, { computed } from '@ember/object';
 
 export default Component.extend({
   store: service(),
+  filterValue: "Title",
+
+  prevDisabled: computed('model.query.page', function() {
+    return ((this.get('model.query.page') - 1) < 0);
+  }),
```
Comment (juandsierra82) [inline app/components/books-deck.js]: ( )??

File: app/components/books-deck.js [hunk 3]
```diff
@@ -1,29 +1,54 @@
...
+  nextDisabled: computed('metaData', function() {
+    return ((this.get('model.query.page') + 1) > (this.get('metaData.total-pages') - 1));
+  }),
+  currentPage: computed('model.query.page', function() {
```
Comment (juandsierra82) [inline app/components/books-deck.js]: nextDisabled and prevDisabled can be one property

File: app/components/books-deck.js [hunk 4]
```diff
@@ -1,29 +1,54 @@
...
+    return (this.get('model.query.page') + 1);
+  }),
+  currentFilterValue: computed('model.query.filterValue', function() {
+    if (this.get('model.query.filterValue') == "name") {
+      return "Title";
```
Comment (juandsierra82) [inline app/components/books-deck.js]: in javascript convention is === not ==. look up type comparisons.

File: app/components/books-deck.js [hunk 5]
```diff
@@ -1,29 +1,54 @@
...
     },
-    //currently under development
-    deleteFromUserLibrary: function(book) {
-      const currentTransaction = this.get('store').peekRecord('transaction', {
-        user: this.get('model.user'),
-        book: book
-      });
-      this.sendAction('deleteFromUserLibrary', currentTransaction.deleteRecord());
+    nextPage() {
+      this.sendAction('nextPage');
```
Comment (juandsierra82) [inline app/components/books-deck.js]: nextPage and prevPage can be one action, and can send one action

File: app/components/books-deck.js [hunk 6]
```diff
@@ -1,29 +1,54 @@
...
+    },
+    prevPage() {
+      this.sendAction('prevPage');
+    },
+    filterBy(param) {
+      this.sendAction('filterBy', param);
+    },
+
```
Comment (juandsierra82) [inline app/components/books-deck.js:46]: filterBy should work using two way data binding for cleaner code.

File: app/components/books-deck.js [hunk 7]
```diff
@@ -1,29 +1,54 @@
...
+    if (this.get('model.query.filterValue') == "name") {
+      return "Title";
+    } else {
+      return "Author";
+    }
+  }),
+
   actions: {
-    setModalBook: function(currentBook) {
+    setModalBook(currentBook) {
```
Comment (juandsierra82) [inline app/components/books-deck.js:27]: think about names

File: app/components/books-deck.js [hunk 8]
```diff
@@ -1,29 +1,54 @@
...
+    prevPage() {
+      this.sendAction('prevPage');
+    },
+    filterBy(param) {
+      this.sendAction('filterBy', param);
+    },
+
+    filterParam(filterProperty) {
+      this.sendAction('filterParam', filterProperty);
+
```
Comment (juandsierra82) [inline app/components/books-deck.js:49]: naming again actions should be action words.

File: app/components/nav-bars/top-navbar.js [hunk 9]
```diff
@@ -5,6 +5,6 @@ export default Component.extend({
 
   tagName: 'nav',
   elementId: 'top-navbar',
-  classNames: ['navbar navbar-dark bg-dark navbar-expand-xl'],
+  classNames: ['navbar navbar-dark bg-dark navbar-expand-xl']
```
Comment (juandsierra82) [inline app/components/nav-bars/top-navbar.js:8]: separate classNames generally; think about use of the variables.

File: app/controllers/books.js [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+  queryParams: ['page', 'currentSearch', 'filterValue'],
+  page: 0,
+  lastPage: 0,
+  filterValue: "name",
+  currentSearch: '',
+
+
+  metaData: computed('model', function() {
+    let meta = this.get('model.meta');
+    return meta;
```
Comment (juandsierra82) [inline app/controllers/books.js]: never use two lines of code when one will do

File: app/controllers/books.js [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+      if (this.get('page') < this.get('metaData.total-pages') - 1) {
+        let page = this.get('page');
+        this.set('page', page + 1);
+      }
+    },
+    prevPage() {
+      if (this.get('page') > 0) {
+        this.set('page', this.get('page') - 1);
+      }
+    },
```
Comment (juandsierra82) [inline app/controllers/books.js]: could be one action to be clearer

File: app/controllers/books.js [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+      param = param.trim();
+      let filterParam = this.get('filterValue');
+      let currentPage = this.get('model.query.page');
+      let lastPage = this.get('lastPage');
+      if (currentPage != 0) {
+        this.set('lastPage', currentPage);
+      }
+      if (param !== '') {
+        if (filterParam == "name") {
+          this.set('page', 0);
```
Comment (juandsierra82) [inline app/controllers/books.js]: see above about ==

File: app/routes/books.js [hunk 13]
```diff
@@ -1,14 +1,29 @@
...
-  model() {
-    return this.store.findAll('book');
+  model(params) {
+    return this.get('store').query('book', params);
   },
   setupController(controller, model) {
     this._super(controller, model);
     controller.set('user', this.modelFor('application').get('user'));
     this.modelFor('application').set('loginHref', false);
+
```
Comment (juandsierra82) [inline app/routes/books.js:29]: if we have a controller/book.js why do we need setupController

File: app/routes/books.js [hunk 14]
```diff
@@ -19,9 +34,8 @@ export default Route.extend({
         this.transitionTo('users.current', user.get('id'));
       })
     },
-    //currently under development
-    deleteFromUserLibrary(transaction) {
-      transaction.save();
+    handleFilterEntry(filterInputValue, filterAction) {
+      filterAction(filterInputValue)
```
Comment (juandsierra82) [inline app/routes/books.js]: delete

File: app/styles/site-layout.scss [hunk 15]
```diff
@@ -16,15 +16,27 @@
...
-      margin-bottom: 1.1rem;
+      margin-bottom: 1rem;
       border-color: black;
       cursor: pointer;
     }
 
     .main-page-book-body {
-      height: 16rem;
+      height: 19rem;
+    }
```
Comment (juandsierra82) [inline app/styles/site-layout.scss]: scss allows nesting

**Review by juandsierra82 — CHANGES_REQUESTED** [review]

File: app/controllers/books.js [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+          this.set('page', this.get('page') - 1);
+        }
+      }
+    },
+    filterBy(param) {
+      param = param.trim();
+      let filterParam = this.get('filterValue');
+      let currentPage = this.get('model.query.page');
+      let lastPage = this.get('lastPage');
+      if (currentPage != 0) {
```
Comment (fangtingliu) [inline app/controllers/books.js]: follow the airbnb guide please.

Comment (fangtingliu) [reply app/controllers/books.js]: '!=‘ should not be used unless it is absolutely necessary

File: app/routes/books.js (hunk 13 above)
Comment (fangtingliu) [reply app/routes/books.js:29]: exactly. why?

**Pull request merged**